fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/PlannerExpressionPointerSet.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * PlannerExpressionPointerSet.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.temp;
+
+import javax.annotation.Nonnull;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+/**
+ * A set of {@link PlannerExpression}s that uses reference ("pointer") equality to determine equivalence for the
+ * purposes of set membership, rather than the {@link #equals(Object)} method used by the Java {@link Set} interface..
+ * This is important for implementing the memo data structure in {@link GroupExpressionRef}
+ * @param <T> the planner expression type contained in the set
+ */
+public class PlannerExpressionPointerSet<T extends PlannerExpression> implements Iterable<T> {
```

Comment:
I think so, but we'd need to make `Wrapper` be `Wrapper<T>` instead of `Wrapper<T extends PlannerExpression>`. I'm not sure which one is actually preferable.

Reply:
Oh, I see. I then it would need to be something like:

```java
if (obj instanceof PlannerExpression) {
    return false;
} else {
   return members.contains(new Wrapper<PlannerExpression>((PlannerExpression)obj));
```

The set itself may not be of type `Wrapper<PlannerExpression>`, but I think that's fine.

Reply:
Okay, this turned out not to be too bad after all, especially once I learned about `AbstractCollection`. Let me know what you think!

Reply:
Ah, nice! Looks good!